torch/_inductor/fx_utils.py

```diff
@@ -75,23 +91,53 @@ class FakeTensorUpdater:
     new hash, and recompute the faketensor metadata for that node. Then, we
     continue to recursively compute the faketensors for all users until the
     fake tensors stop changing.
+
+    Since this runs in the context of Inductor, we assume that the input and
+    output semantics for the graph (and any subgraphs) are not subject to change
+    after class initialization.  Any violations of this assumption may result in
+    undefined behavior.
     """
 
-    def __init__(self, graph: torch.fx.Graph) -> None:
-        self.processed_hashes = OrderedSet[Any]()
-        self.graph = graph
+    def __init__(self, gm: torch.fx.GraphModule) -> None:
+        self.processed_hashes = OrderedSet[_FxNodeHash]()
+        self.gm = gm
+
+        # Import here to avoid circular import issues.
+        from torch._inductor.compile_fx import _get_subgraph_names
 
-        for node in self.graph.nodes:
+        self.subgraph_updaters = {
+            name: FakeTensorUpdater(getattr(self.gm, name))
+            for name in _get_subgraph_names(self.gm)
+        }
+
+        for node in self.gm.graph.nodes:
             self.processed_hashes.add(self.hash_node(node))
 
-    def hash_node(self, node: torch.fx.Node):
-        # todo(chilli): Not a great hash function
-        return (node, node.target, id(node.args), id(node.kwargs))
+    def _is_subgraph_node(self, node: Any) -> bool:
+        return (
+            isinstance(node, torch.fx.Node)
+            and node.op == "get_attr"
+            and node.target in self.subgraph_updaters
+        )
+
+    def hash_node(self, node: torch.fx.Node) -> _FxNodeHash:
+        if self._is_subgraph_node(node):
+            assert isinstance(node.target, str)
+            return _FxNodeHash(
+                node,
+                node.target,
+                id(node.args),
+                id(node.kwargs),
+                _FakeTensorUpdaterHash(
```

Comment:
Hash of a subgraph node includes the hashes of all of the nodes in the subgraph? (If I'm reading this correctly?)

Reply:
When the `FakeTensorUpdater` class is initialized, it automatically adds hashes for every node present in the graph at initialization, so this contains all hashes, not just updated ones.